ironfish/src/syncer.ts

```diff
@@ -99,46 +111,171 @@ export class Syncer {
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    let candidateCheckCount = 0
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      // We only want to successfully measure so many candidates per measurement
+      // phase
+      if (candidateCheckCount > 8) {
+        break
+      }
+
+      const peerIdentity = peer.getIdentity()
+      if (peerIdentity == null) {
+        continue
+      }
+
+      const start = BenchUtils.start()
+      try {
+        const response = await this.peerNetwork.getBlockHeaders(peer, 1, 1)
+        if (response.headers.length !== 1) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong number of block headers`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+        if (!response.headers[0].hash.equals(this.chain.genesis.hash)) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong block header`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+      } catch (e) {
+        this.logger.debug(
+          `Error while trying to measure peer '${
+            peer.displayName
+          }', skipping this peer: ${ErrorUtils.renderError(e)}`,
+        )
+        continue
+      }
+      const rtt = BenchUtils.end(start)
+      peerRtt.set(peerIdentity, rtt)
+      candidateCheckCount += 1
+    }
+
+    // Sort the peers by the round-trip-time of the block header request and get
+    // the fastest one to sync from
+    const fastestCandidateIdentity = [...peerRtt.entries()].sort((a, b) => a[1] - b[1])[0][0]
+    if (!fastestCandidateIdentity) {
+      return
+    }
```

Comment:
I don't think we need this check. I feel like a better one might be to add a `if (peerRtt.size() < 1) return` before the sort

Reply:
> I don't think we need this check.

The TS compiler doesn't seem to pick up on this, but it _is_ possible for the set to be empty.

> I feel like a better one might be to add a if (peerRtt.size() < 1) return before the sort

Better how? :D

Reply:
If the set is empty, wouldn't the [0][0] throw an exception?

Reply:
Actually, I think this will throw an Error. TS makes nested arrays kinda easy to cause problems. Try in the Node.js repl:

```
const set = new Set()
[...set.entries()].sort((a, b) => a[1] - b[1])[0][0]
Uncaught TypeError: Cannot read properties of undefined (reading '0')
```
You could do this to preserve the original intent:
```
[...set.entries()].sort((a, b) => a[1] - b[1])[0]?.[0]
```